### ISSUE TITLE
App records not recognised

### ISSUE THREAD

**Issue by davidnicholls125** [johnvanbreda/naturespot-d8#117]
My understanding is that records submitted via the app (linked to NS) are initially recorded on the iRecord species list then mapped onto the NS list - unless we don't have that species when it is left unchanged. However I'm seeing a number of app records that are still on the the iRecord inventory even though we have that species on our list. Example are the birds below and also Grey Squirrel - Sciurus carolinensis (appearing under terrestrial mammals). Can you see why?
![image](https://user-images.githubusercontent.com/80914584/190863332-e4cede05-b91e-45d4-a0e3-80c6852b7e88.png)

**Comment by johnvanbreda** [comment]
@davidnicholls125 what makes you think that they are still on the iRecord/UKSI list? I checked the first 2 records above and they are both linked to the NatureSpot list.

**Comment by davidnicholls125** [comment]
They are appearing on the elasticsearch verification page under 'bird' and not 'Birds' as we use. When this happens they get missed by the verifier filtering using the NS taxon group names - particularly with insect groups.

**Comment by johnvanbreda** [comment]
Am I right in thinking you've manually corrected these records now? If you see the problem again please can you leave one record as it is, then let me know the ID?

It's odd because I can see the script to move the records has worked because the website ID changed, but it failed to update the taxon. If I run a test script for the Whitethroat record it proves that the mapping seems to work OK, so I can't really see why it might have failed.

**Comment by davidnicholls125** [comment]
Yes I corrected the birds. Look at Sciurus carolinensis below. These are listed under 'terrestrial mammals' but with the same taxon name we have on NS.
![image](https://user-images.githubusercontent.com/80914584/191502361-fa9ae187-0df3-4da7-a28a-1ed59ca087cb.png)

**Comment by johnvanbreda** [comment]
I've not been able to spot the cause, but I have written a script (below) which updates any that haven't been mapped across. It fixed 255 occurrences.

```sql

drop table if exists to_transfer_occurrences;

select distinct o.id as occurrence_id, o.taxa_taxon_list_id
into temporary to_transfer_occurrences
from occurrences o
join taxa_taxon_lists ttl on ttl.id=o.taxa_taxon_list_id and ttl.taxon_list_id=15
where o.website_id=8;

-- map taxonomy where possible
drop table if exists mappings;

select distinct taxa_taxon_list_id as old_uksi_id, null::integer as new_ns_id
into temporary mappings
from to_transfer_occurrences;

update mappings m
set new_ns_id=cttl2.id
from cache_taxa_taxon_lists cttl1
join cache_taxa_taxon_lists cttl2 on cttl2.external_key=cttl1.external_key
  and cttl2.taxon_list_id=8
  -- Exact name matches will be OK
  and lower(regexp_replace(regexp_replace(replace(replace(cttl2.taxon, '''', ''), '-', ' '), E'\\(.+\\)', '', 'g'), '\s+', ' ', 'g'))
    = lower(regexp_replace(regexp_replace(replace(replace(cttl1.taxon, '''', ''), '-', ' '), E'\\(.+\\)', '', 'g'), '\s+', ' ', 'g'))
where cttl1.id=m.old_uksi_id
and cttl1.taxon_list_id=15;

update occurrences o
set taxa_taxon_list_id=m.new_ns_id
from to_transfer_occurrences tto, mappings m
where o.id=tto.occurrence_id
and o.taxa_taxon_list_id=m.old_uksi_id
and m.new_ns_id is not null;

insert into work_queue(task, entity, record_id, cost_estimate, priority, created_on)
select distinct 'task_cache_builder_update', 'occurrence', tto.occurrence_id, 100, 2, now()
from to_transfer_occurrences tto
join occurrences o on o.id=tto.occurrence_id
join cache_taxa_taxon_lists cttl on cttl.id=o.taxa_taxon_list_id and cttl.taxon_list_id=8
left join work_queue q on q.task='task_cache_builder_update' and q.entity='occurrence' and q.record_id=tto.occurrence_id
where q.id is null;
```

**Comment by davidnicholls125** [comment]
Will this script have to be manually run on a regular basis? Can it be automated?

**Comment by johnvanbreda** [comment]
I just tried the script again and there were no updates required. So, I hope that whatever happened was a temporary thing and the script does not need to be run. I suggest leaving this issue open for a while, then trying the script again in future. If there updates then I will integrate this into the background task that moves app records, but if there is nothing to do we can close the issue.

**Comment by johnvanbreda** [comment]
I just ran the script and there were 3 records to update, but I think probably due to taxon amendments (since if there was a real problem you'd expect more). Will leave this open and try again in future, but will look in detail at any amendments to see if I'm right.
